File: util/android/src/main/java/com/dmi/util/opengl/GLSurfaceViewExt.java
package com.dmi.util.opengl;

import android.content.Context;
import android.opengl.GLSurfaceView;
import android.util.AttributeSet;

import java.util.concurrent.atomic.AtomicBoolean;

import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.opengles.GL10;

public class GLSurfaceViewExt extends GLSurfaceView {
    private GLRenderer renderer;

    private final AtomicBoolean renderRun = new AtomicBoolean(false);
    private boolean needFreeResources = false;

    public GLSurfaceViewExt(Context context) {
        super(context);
    }

    public GLSurfaceViewExt(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    public void onPause() {
        synchronized (renderRun) {
            if (renderRun.get()) {
                queueEvent(this::freeResources);
                super.onPause();
            }
        }
    }

    @Override
    public void onResume() {
        synchronized (renderRun) {
            if (renderRun.get()) {
                super.onResume();
            }
        }
    }

    @Override
    protected void onDetachedFromWindow() {
        queueEvent(this::freeResources);
        super.onDetachedFromWindow();
    }

    @Override
    public void requestRender() {
        synchronized (renderRun) {
            if (renderRun.get()) {
                super.requestRender();
            }
        }
    }

    public void setRenderer(GLRenderer renderer) {
        this.renderer = renderer;
        synchronized (renderRun) {
            super.setRenderer(new Renderer() {
                @Override
                public void onSurfaceCreated(GL10 gl, EGLConfig config) {
                    needFreeResources = true;
                    renderer.onSurfaceCreated();
                }

                @Override
                public void onSurfaceChanged(GL10 gl, int width, int height) {
                    renderer.onSurfaceChanged(width, height);
                }

                @Override
                public void onDrawFrame(GL10 gl) {
                    renderer.onDrawFrame();
                }
            });
            renderRun.set(true);
        }
    }

    private void freeResources() {
        if (needFreeResources) {
            renderer.onFreeResources();
            needFreeResources = false;
        }
    }

    @Override
    public void setRenderer(Renderer renderer) {
        throw new RuntimeException("You should use setRenderer(GLRenderer)");
    }
}

File: util/util-android/src/main/kotlin/com/dmi/util/opengl/GLSurfaceViewExt.kt
package com.dmi.util.opengl

import android.content.Context
import android.opengl.GLSurfaceView
import android.util.AttributeSet

import java.util.concurrent.atomic.AtomicBoolean

import javax.microedition.khronos.egl.EGLConfig
import javax.microedition.khronos.opengles.GL10

open class GLSurfaceViewExt : GLSurfaceView {
    private var renderer: GLRenderer? = null

    private val renderRun = AtomicBoolean(false)

    constructor(context: Context) : super(context) {
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
    }

    override fun onPause() {
        synchronized (renderRun) {
            if (renderRun.get()) {
                super.onPause()
            }
        }
    }

    override fun onResume() {
        synchronized (renderRun) {
            if (renderRun.get()) {
                super.onResume()
            }
        }
    }

    override fun requestRender() {
        synchronized (renderRun) {
            if (renderRun.get()) {
                super.requestRender()
            }
        }
    }

    fun setRenderer(renderer: GLRenderer) {
        this.renderer = renderer
        synchronized (renderRun) {
            super.setRenderer(object : GLSurfaceView.Renderer {
                override fun onSurfaceCreated(gl: GL10, config: EGLConfig) {
                    renderer.onSurfaceCreated()
                }

                override fun onSurfaceChanged(gl: GL10, width: Int, height: Int) {
                    renderer.onSurfaceChanged(width, height)
                }

                override fun onDrawFrame(gl: GL10) {
                    renderer.onDrawFrame()
                }
            })
            renderRun.set(true)
        }
    }

    override fun setRenderer(renderer: GLSurfaceView.Renderer) {
        throw RuntimeException("You should use setRenderer(GLRenderer)")
    }
}